---
title: 'SPADE: a spatial pattern and differential expression analysis method with spatial transcriptomic data'
author: "Fei Qin, Feifei Xiao, Guoshuai Cai"
date: 'Last updated: 04/03/2023'
output:
  pdf_document: default
  html_document:
    df_print: paged
packages: SPADE
fontsize: 12pt
---

# 1. Introduction to the SPADE method

To identify spatially variable (SV) genes with spatial transcriptomic data, the SPADE method was developed based on a Gaussian process regression (GPR) model with Gaussian kernel. GPR model can model the relationship between gene expression and other covariates (i.e., cell groups) incorporating the spatial information of multiple cells. Besides detecting SV genes within groups, SPADE also provided a framework to identify SV genes between different treatment conditions. The framework of the SPADE method is summarized and illustrated in Figure below. First, original read counts data were normalized into continuous data using a two-step normalization strategy. Second, instead of using a fixed length scale hyperparameter across genes in the kernel function of GPR model, SPADE estimated the optimal hyperparameter for each gene to improve the accuracy of SV gene identification. To identify SV genes within groups, hypothesis testing was conducted based on a quadratic score statistic with a Davies method to compute the P value. With SV gene detection between groups, SPADE exchanged the optimal hyperparameters estimated in two groups and  utilized a crossed likelihood-ratio test to calculate the P value for each gene. 

![FLCNA framework](E:/DB/Dropbox/Qinfei/Research/Spatial scRNA/Code/SPADE_package/vignettes/Figure 1.png)

# 2. Installation

```{r install-bioc, message=FALSE, warning = FALSE, eval=FALSE}
library(devtools)
install_github("thecailab/SPADE")
```





# 3. Identifying SV genes within groups

To characterize complex tissues with spatial transcriptomics, the identification of SV genes within groups is an essential first step. These identified SV genes can be defined as genes with uneven, aggregated, or patterned spatial distribution of gene expression magnitudes, and they play important role in biomarker discovery and drug development.

# 3.1 Data
To help illustrate how SPADE package can be applied, the SeqFISH dataset was provided in the package. The SeqFISH dataset was collected on the mouse hippocampus with 249 genes measured on 257 spots. After filtering out cells located at the boundary to relieve border artifacts, a final set of 249 genes measured on 131 spots were retained for the analysis. Two data files were given in the package including read counts data and coordinates information data. The application of SPADE includes steps of normalization, parameters estimation and testing.

```{r, message=FALSE, warning = FALSE}
library(SPADE)
data(SeqFISH)
data(info)
readcounts <- SeqFISH
```

```{r, message=FALSE, warning = FALSE}
dim(readcounts)
dim(info)
```

```{r, message=FALSE, warning = FALSE}
readcounts[1:5,1:5]
head(info)
```


   
## 3.2 Normalization

A two-step normalization strategy was implemented to transformed original read counts data into continuous data. First, the mean-variance dependency was stabilized using an Anscombeâ€™s transformation strategy. Second, a linear regression was utilized to regress out library size from the above transformed expression data. After this step, the normalized continuous data will be utilized for the parameter estimation and hypothesis testing in the SPADE method. SPADE_norm() R function is used for the normalization. 

```{r, message=FALSE, warning = FALSE}
D2_norm <- SPADE_norm(readcounts=as.matrix(D2_data), info=D2_info)
D5_norm <- SPADE_norm(readcounts=as.matrix(D5_data), info=D5_info)
```

## 3.3 Parameter estimation 

The GPR model was used to model the expression of each gene across cells with different locations. Theoretically, the problem of finding SV genes in the SPADE method is to test how well the candidate covariance matrix fits the spatial transcriptomic data. For each gene, we estimated the optimal length-scale hyperparameter in the Gaussian kernel to increase the accuracy of SV gene identification. SPADE_estimate() R function is used for the parameter estimation. 

```{r, message=FALSE, warning = FALSE, results='hide'}
Est <- SPADE_estimate(expr_data=data_norm, info=info)

```

```{r, message=FALSE, warning = FALSE}
head(Est)
```

The output of SPADE_estimate() includes 1) Gene ID; 2) theta_Gau: Optimal length-scale hyperparameter in kernel function estimated for each gene. 3) Lik_Gau: Maximum likelihood computed for each gene. 

## 3.4 Testing

After the optimal length-scale hyperparameter was estimated, P-value for each gene was computed based on a quadratic score statistic with a Davies method. SPADE_test() is used for computing P value for each gene.

```{r, message=FALSE, warning = FALSE, results='hide'}
Test_res <- SPADE_test(object=data_norm, location=info, para=Est)
```


```{r, message=FALSE, warning = FALSE}
Test_res[c(216, 230),]
sum(Test_res$Adjust.Pvalue < 0.05)
```

```{r, message=FALSE, warning = FALSE, echo=FALSE}
source("E:\\DB\\Dropbox\\Qinfei\\Research\\Spatial scRNA\\Code\\SPADE_package\\utilities.R")
# sp_sim_count    <- as(as.matrix(readcounts),"sparseMatrix")

library(ggplot2)
pltdat  <- cbind.data.frame(info,apply(as.matrix(readcounts),1,relative_func))
pp      <- lapply(c(216,230),function(x){pattern_plot_SPADE(pltdat,x,pointsize=3,main=T, titlesize=1,
                                                               min.pand=0.8,max.pand=1.01,
                                                               title=substr(rownames(readcounts),3,nchar(rownames(readcounts))-1)[x],
                                                               pal=colorRampPalette(c("grey95","grey50","deeppink1","deeppink3")))})
Fig_seqFISH    <- ggpubr::ggarrange(pp[[1]], pp[[2]], 
                            common.legend = TRUE, heights = c(2, 2), legend="right", align = "h")
Fig_seqFISH  
```

The output of SPADE_test() includes 1) Gene name; 2) Q value: a score statistic used to calculate P value. Q follows a mixture of independent chi-square distributions with mixing weights that depend on the eigenvalues of the kernel matrix. 3) P value: Based on a chi-square mixture distribution and Q value, an exact method based on the Davies method is utilized to compute P-values for genes to define SV genes. 4) Adjusted P values: P-values across all genes are adjusted with the Benjamini and Hochberg method to correct for occurrence of false positives. 






# 4. Identifying SV genes between groups

SPADE was the first method to investigate SV genes between groups with spatial transcriptomic data. The identification of SV genes between groups was essential for understanding the changes of spatial patterns with different treatment conditions or different time phases. 

## 4.1 Data

To illustrate how SPADE can be applied to identify SV genes between groups. A real spatial transcriptomic dataset with axolotl telencephalon (i.e., ARRISTA) was provided in the package. Two different post injury stages (i.e., 2DPI, 5DPI) were included in the dataset. This data contain three genes with 1,188 spots and 938 spots in the 2DPI group and 5DPI group, respectively.

```{r, message=FALSE, warning = FALSE}
data(D2_data)
data(D2_info)
data(D5_data)
data(D5_info)
dim(D2_data)
dim(D2_info)
dim(D5_data)
dim(D5_info)
```

## 5.1 Normalization

Still, the same two-step normalization strategy was implemented for read counts data from each group to transform original data into continuous data.

```{r, message=FALSE, warning = FALSE}
D2_norm <- SPADE_norm(readcounts=as.matrix(D2_data), info=D2_info)
D5_norm <- SPADE_norm(readcounts=as.matrix(D5_data), info=D5_info)
```

## 5.2 Parameter estimation and testing

SPADE can also identify SV genes between groups based on a crossed likelihood ratio test with spatial transcriptomic data. SPADE first estimate the optimal hyperparameter for kernel matrix in each group, respectively. Thus, for each gene, the log likelihood in each group can be easily calculated with its optimal kernel. Then we exchange the estimated hyperparameters to compute the log likelihoods for both groups, and compare them to their optimal log likelihoods. The likelihood ratio test statistic is utilized to identify SV genes with P-values calculated using F test with degree freedom of one. 

```{r, message=FALSE, warning = FALSE}
res <- SPADE_DE(D2_norm, D5_norm, D2_info, D5_info) 
res
```


```{r, message=FALSE, warning = FALSE, echo=FALSE}
source("E:\\DB\\Dropbox\\Qinfei\\Research\\Spatial scRNA\\Code\\SPADE_package\\utilities.R")
library(ggplot2)
pltdat2  <- cbind.data.frame(D2_info, apply(D2_data, 1, relative_func))
colnames(pltdat2)[1:2] <- c("x","y")
pp2      <- lapply(c(1:3),function(x){pattern_plot_SPADE(pltdat2, x, pointsize=3, main=F, titlesize=1,
                                                            min.pand=0.8, max.pand=1.01, title="2DPI",
                                                            pal=colorRampPalette(c("grey95", "grey50", "deeppink1", "deeppink3")))})



pltdat5  <- cbind.data.frame(D5_info, apply(D5_data, 1, relative_func))
colnames(pltdat5)[1:2] <- c("x","y")
pp5      <- lapply(c(1:3),function(x){pattern_plot_SPADE(pltdat5, x, pointsize=3 ,main=F, titlesize=1,
                                                         min.pand=0.8, max.pand=1.01, title="5DPI",
                                                         pal=colorRampPalette(c("grey95","grey50","deeppink1","deeppink3")))})

Fig_ARRISTA    <- ggpubr::ggarrange(pp2[[1]], pp2[[2]], pp2[[3]], 
                                    pp5[[1]], pp5[[2]], pp5[[3]],
                                    common.legend = TRUE, legend="right", nrow = 2, ncol=3)
Fig_ARRISTA  
```