---
title: 'SPADE: a spatial pattern investigation method, to identify spatially expressed genes within groups and between 
groups using spatial transcriptomics data.'
author: "Fei Qin, Feifei Xiao, Guoshuai Cai"
date: 'Last updated: 03/23/2023'
output:
  pdf_document: default
  html_document:
    df_print: paged
packages: SPADE
---

# 1. Introduction to the SPADE method

To identify SE genes with spatially resolved transcriptomics data, the SPADE method was developed based on a GPR model, which can model the relationship between gene expression and other covariates (i.e., cell groups) incorporating the spatial information of multiple cells. SPADE was able to identify SE genes within a single group and between groups. The framework of the SPADE method is summarized and illustrated in Figure 1. First, original read counts data were normalized into continuous data. Second, instead of using a fixed length scale hyperparameter in the covariance kernel of GPR, SPADE estimated the optimal hyperparameter for each gene. To identify SE genes within groups, hypothesis testing was conducted based on a quadratic score statistic [13] with a Davies method [14] to compute the P value. With SE gene detection between groups, SPADE exchanged the optimal hyperparameters estimated in two groups and then utilized a crossed likelihood ratio test to calculate the P value for each gene. 

![FLCNA framework](E:/DB/Dropbox/Qinfei/Research/Spatial scRNA/Code/SPADE_package/vignettes/Figure 1.png)

# 2. Installation

```{r install-bioc, message=FALSE, warning = FALSE, eval=FALSE}
library(devtools)
install_github("thecailab/SCRIP")
```

# 3. Data
To help explain 

# 4. Identifying SE genes within groups

## 4.1 Normalization

## 4.2 Parameter estimation 

## 4.3 Testing


# 5. Identifying SE genes between groups

## 5.1 Normalization

## 5.2 Parameter estimation and testing


A two-step median normalization approach is implemented to remove the effect of biases from the GC-content and mappability. We further calculate the ratio of normalized RC and its sample specific mean, and the logarithm transformation of this ratio (log2R) is used in the main step of the FLCNA method. FLCNA_normalization() R function is used for the normalization. 

```{r, message=TRUE, warning = FALSE}
log2Rdata <- FLCNA_normalization(Y=QCobject$Y, gc=QCobject$ref$gc, map=QCobject$ref$mapp)
```

# 4.2 Simultaneous CNA detection and subclone clustering

Subclone clustering is achieved based on a GMM, and breakpoints detection is conducted by adding a fused lasso penalty term to the typical GMM model. FLCNA() R function can be used for the CNA detection and simultaneous subclone clustering. There  are two hyperparameters to be pre-defined in the FLCNA method, including the number of clusters K and the tuning parameter lambda. The tuning hyperparameter lambda is used to control the overall number of change points that less change points tend to be generated with larger lambda value. To find the optimal values of K and lambda, we use a BIC-type criterion, and the model with smallest BIC value is selected as the optimal model. 

```{r, message=TRUE, warning = FALSE, eval=FALSE}
# K: The number of clusters.
# lambda: The tuning parameter in the penalty term, the default is 3.
output_FLCNA <- FLCNA(K=c(3,4,5), lambda=3, Y=log2Rdata)
```

```{r, message=FALSE, warning = FALSE, echo=FALSE}
data(output_FLCNA)
```

```{r, message=TRUE, warning = FALSE}
# The number of clusters in the optimal model
output_FLCNA$K.best
```

```{r, message=TRUE, warning = FALSE}
# The estimated mean matrix for K clusters
output_FLCNA$mu.hat.best[,1:11]
```

```{r, message=TRUE, warning = FALSE}
# The cluster index for each cell
output_FLCNA$s.hat.best
table(output_FLCNA$s.hat.best)
```

